llvm/lib/Target/Mips/MipsAsmPrinter.cpp

```diff
@@ -249,6 +254,55 @@ void MipsAsmPrinter::emitJumpTableDest(MCStreamer &OutStreamer,
   EmitToStreamer(OutStreamer, LoadI);
 }
 
+bool MipsAsmPrinter::tryEmitImm48Fix(MCStreamer &OutStreamer,
+                                     const MachineInstr *MI) {
+  if (!FixImm48)
+    return false;
+
+  if (!Subtarget->hasFixImm48())
+    return false;
+
+  if (MI->getOpcode() != Mips::Li_NM &&
+      MI->getOpcode() != Mips::ADDIU48_NM)
+    return false;
+
+  int LastOp = MI->getNumOperands() - 1;
+  auto Last = MI->getOperand(LastOp);
+
+  if (!Last.isImm())
+    return false;
+
+  int64_t Value = Last.getImm();
+
+  auto ImmValueNeedsImm48Fix = [MI](int64_t Value) -> bool {
+    if ((MI->getOpcode() == Mips::Li_NM) &&
+        (isUInt<16>(Value) || isInt<9>(Value) || ((Value && 0xfffu) == 0u)))
```

Comment:
I guess this is excluding the smaller forms of Li to leave only the 48-bit forms? It seems odd to mix `is*Int` with an explicit mask.

If the last term is an explicit mask? If so, typo: `&& -> &`, but it's also redundant because anything that matches is already matched by `isUint<16>`.

Reply:
I realise now I read the sense of the masked test of `0xfff` incorrectly. So I guess this is intended to exclude patterns that will be matched as `LUI` by the assembler?